File: hack/merge_cluster_roles.py
#!/usr/bin/python
""" merge_cluster_roles.py - merge OpenShift cluster roles into one """
# Copyright (c) 2018 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#  http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import unicode_literals, print_function
import os.path
import sys
import yaml
import warnings


def main():
    base = ClusterRole()
    sources = [os.path.relpath(sys.argv[1])]
    with open(sys.argv[1], 'r') as f:
        base = ClusterRole(yaml.load(f, Loader=yaml.SafeLoader))

    manifests = sys.argv[2:]
    for manifest in sorted(manifests):
        with open(manifest, 'r') as f:
            y = yaml.load(f, Loader=yaml.SafeLoader)
            if y['kind'].endswith('RoleList'):
                for item in y['items']:
                    base.merge(ClusterRole(item))
            elif y['kind'].endswith('Role'):
                base.merge(ClusterRole(y))
            else:
                warnings.warn('unexpected resource kind: {}'.format(y['kind']))
                continue
        sources.append(os.path.relpath(manifest))

    print('---')
    print('# This is a generated file. DO NOT EDIT')
    print('# Run `make merge-cluster-roles` to generate.')
    print('# Sources: ')
    for source in sources:
        print('# \t' + source)
    print(yaml.dump(base.manifest, default_flow_style=False))


class ClusterRole(object):

    def __init__(self, manifest = {}):
        self.manifest = manifest
        # In theory this merges ClusterRoles.
        # If it happens to merge a regular Role, then log a warning
        # but do not raise an exception, as this behavior is needed.
        if 'kind' in self.manifest and self.manifest['kind'] != 'ClusterRole':
            warnings.warn('creating a ClusterRole from a {}'.format(self.manifest['kind']))
        self.rules = {}
        if 'rules' in self.manifest:
            for r in self.manifest['rules']:
                self.rules[rule_to_string(r)] = r
        else:
            self.manifest['rules'] = []

    # merge adds the rules of the passed ClusterRole to the list of rules.
    def merge(self, cr):
        for s, r in cr.rules.items():
            if s not in self.rules:
                self.rules[s] = r
                self.manifest['rules'].append(r)


# rule_to_string stably stringifies a rule so that rules can be
# deduplicated using a dict.
def rule_to_string(rule):
    ks = list(rule.keys())
    ks.sort()
    s = []
    for k in ks:
        v = rule[k]
        if isinstance(v, list):
            v.sort()
        s.append('{}:{}'.format(k, v))
    return ','.join(s)


if __name__ == "__main__":
    main()

File: hack/merge_cluster_roles.go
// Copyright 2020 The Cluster Monitoring Operator Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/ghodss/yaml"

	rbacv1 "k8s.io/api/rbac/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type ClusterRole struct {
	rbacv1.ClusterRole `yaml:",inline"`
	rules              map[string]rbacv1.PolicyRule
}

func NewClusterRole(manifest []byte) *ClusterRole {
	var (
		meta metav1.TypeMeta
		cr   ClusterRole
	)

	err := yaml.Unmarshal(manifest, &meta)
	if err != nil {
		log.Fatal(err)
	}

	switch kind := meta.Kind; {
	case strings.HasSuffix(kind, "RoleList"):
		var crList rbacv1.ClusterRoleList
		err = yaml.Unmarshal(manifest, &crList)
		if err != nil {
			log.Fatal(err)
		}
		for _, role := range crList.Items {
			cr.merge(&ClusterRole{ClusterRole: role})
		}
	case strings.HasSuffix(kind, "Role"):
		if kind != "ClusterRole" {
			log.Printf("creating a ClusterRole from a %s", kind)
		}
		err = yaml.Unmarshal(manifest, &cr)
		if err != nil {
			log.Fatal(err)
		}
	default:
		log.Printf("unexpected resource kind: %s", meta.Kind)
	}

	cr.rules = make(map[string]rbacv1.PolicyRule)
	for _, rule := range cr.Rules {
		cr.rules[rule.String()] = rule
	}

	return &cr
}

func (cr *ClusterRole) merge(role *ClusterRole) {
	for s, r := range role.rules {
		_, ok := cr.rules[s]
		if !ok {
			cr.rules[s] = r
			cr.Rules = append(cr.Rules, r)
		}
	}
}

func printBoilerplate() {
	fmt.Println(`---
# This is a generated file. DO NOT EDIT
# Run ` + "`make merge-cluster-roles`" + ` to generate.`)
}

func printSources() {
	fmt.Println("# Sources: ")
	for _, source := range os.Args[1:] {
		fmt.Printf("# \t%s\n", source)
	}
}

func printClusterRole(cr *ClusterRole) {
	crManifest, err := yaml.Marshal(&cr)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(string(crManifest))
}

func main() {
	base, err := ioutil.ReadFile(filepath.Clean(os.Args[1]))
	if err != nil {
		log.Fatal(err)
	}
	cr := NewClusterRole(base)

	for _, manifest := range os.Args[2:] {
		data, err := ioutil.ReadFile(filepath.Clean(manifest))
		if err != nil {
			log.Fatal(err)
		}
		cr.merge(NewClusterRole(data))
	}

	// Rules need to be sorted to avoid having an ever changing manifest.
	sort.Slice(cr.Rules, func(i, j int) bool {
		return strings.Compare(cr.Rules[i].String(), cr.Rules[j].String()) < 0
	})

	printBoilerplate()
	printSources()
	printClusterRole(cr)
}
